automation/Extension/firefox/lib/http-post-parser.js

```diff
@@ -0,0 +1,250 @@
+
+var HttpPostParser = function(stream) {
+  // Scriptable Stream Constants
+  this.seekablestream = stream;
+  this.stream = components.classes["@mozilla.org/scriptableinputstream;1"].createInstance(components.interfaces.nsIScriptableInputStream);
+  this.stream.init(this.seekablestream);
+
+  this.postBody = "";
+  this.postLines = [];
+  this.postHeaders = [];
+  // Check if the stream has headers
+  this.hasheaders = false;
+  this.body = 0;
+  if (this.seekablestream instanceof components.interfaces.nsIMIMEInputStream) {
+    this.seekablestream.QueryInterface(components.interfaces.nsIMIMEInputStream);
+    this.hasheaders = true;
+    this.body = -1;
+  } else if (this.seekablestream instanceof components.interfaces.nsIStringInputStream) {
+    this.seekablestream.QueryInterface(components.interfaces.nsIStringInputStream);
+    this.hasheaders = true;
+    this.body = -1;
+  }
+}
+
+HttpPostParser.prototype.rewind = function() {
+  this.seekablestream.seek(0, 0);
+};
+
+HttpPostParser.prototype.tell = function() {
+  return this.seekablestream.tell();
+};
+
+HttpPostParser.prototype.readLine = function() {
+  var line = "";
+  var size = this.stream.available();
+  for (var i = 0; i < size; i++) {
+    var c = this.stream.read(1);
+    if (c == '\r') {
+    } else if (c == '\n') {
+      break;
+    } else {
+      line += c;
+    }
+  }
+  return line;
+};
+
+// visitor can be null, function has side-effect of setting body
+HttpPostParser.prototype.headers = function() {
+  if (this.hasheaders) {
+    this.rewind();
+    var line = this.readLine();
+    while (line) {
+      var keyValue = line.match(/^([^:]+):\s?(.*)/);
+      // match can return null...
+      if (keyValue) {
+        this.postHeaders[keyValue[1]] = keyValue[2];
+      } else {
+        this.postLines.push(line);
+      }
+      line = this.readLine();
+    }
+    this.body = this.tell();
+  }
+};
+
+HttpPostParser.prototype.convertTextPlainToUrlEncoded = function(postBody){
+  /* Convert from text/plain to application/x-www-form-urlencoded
+   * This is to unify the encoding so that we can parse different encodings at one place
+    See, https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest#Using_nothing_but_XMLHttpRequest
+
+    We convert from (text/plain )...
+      foo=bar
+      baz=The first line.
+      The second line.
+    to (application/x-www-form-urlencoded):
+      foo=bar&baz=The+first+line.%0D%0AThe+second+line.%0D%0A
+   */
+  var lines = postBody.split("\n");
+  var post_vars = [];
+  for (var line of lines){
+    if (line.indexOf("=") != -1) {
+      post_vars.push(encodeURIComponent(line.trim()));
+    } else {
+      var x = encodeURIComponent("\r\n" + line.trim());
+      post_vars.push(post_vars.pop() + x);
+    }
+  }
+  return post_vars.join("&");
+}
+
+HttpPostParser.prototype.parseEncodedFormData = function(formData, encodingType){
+  var obj = {};
+
+  if (formData.indexOf("=") == -1)  // not key=value form
```

Comment:
Good catch!

I added a failing test case (`test_record_post_data_ajax_no_key_value_base64_encoded`), fixed it and then [added comments about the issue](https://github.com/gunesacar/OpenWPM/commit/84f1a79122dad0d65f2cb50e65caa9fd63054758#diff-455c9b9d783e44486d3b4a7cb6fe0ec1R149).

Still, there's a risk of "over-parsing". We've some naive checks to infer key=value data. We may work on improving them.

The problem is that sender can craft POST data in any format, including ambiguous forms where it's impossible to tell non-key value data that contains `=` form from genuine form-data.

I checked how Firefox devtools interface handles ambiguous cases, and to my surprise it does over-parse as well:

New test mentioned above sends a base64 string using jQuery's `post` with default arguments:
`$.post("bogus_submit.html", "dGVzdEBleGFtcGxlLmNvbSArIG5hbWUgc3VybmFtZQ==");`

Checking the Params panel for the POST request, `dGVzdEBleGFtcGxlLmNvbSArIG5hbWUgc3VybmFtZQ== ` is parsed as `key:dGVzdEBleGFtcGxlLmNvbSArIG5hbWUgc3VybmFtZQ` and `value: ""` ('=' is perhaps lost due to split or decoding)

![base64_post](https://cloud.githubusercontent.com/assets/5788790/21528915/31f6781e-cd36-11e6-9604-d18d9de0545f.png)